core/src/main/scala/kafka/controller/ControllerChannelManager.scala

```diff
@@ -397,8 +398,25 @@ abstract class AbstractControllerBrokerRequestBatch(config: KafkaConfig,
                                       topicPartition: TopicPartition,
                                       deletePartition: Boolean): Unit = {
     brokerIds.filter(_ >= 0).foreach { brokerId =>
-      val stopReplicaInfos = stopReplicaRequestMap.getOrElseUpdate(brokerId, ListBuffer.empty[StopReplicaRequestInfo])
-      stopReplicaInfos.append(StopReplicaRequestInfo(PartitionAndReplica(topicPartition, brokerId), deletePartition))
+      val result = if (deletePartition)
+        stopReplicaRequestMapWithDelete.getOrElseUpdate(brokerId, mutable.Map.empty)
+      else
+        stopReplicaRequestMapWithoutDelete.getOrElseUpdate(brokerId, mutable.Map.empty)
+
+      // A sentinel (-2) is used as an epoch if the topic is queued for deletion or
+      // does not have a leader yet. This sentinel overrides any existing epoch.
+      val leaderEpoch = if (controllerContext.isTopicQueuedUpForDeletion(topicPartition.topic)) {
+        LeaderAndIsr.EpochDuringDelete
+      } else {
+        controllerContext.partitionLeadershipInfo.get(topicPartition)
+          .map(_.leaderAndIsr.leaderEpoch)
+          .getOrElse(LeaderAndIsr.EpochDuringDelete)
+      }
```

Comment:
👀 This is the logic which handle the sentinel. `-2` is used when a topic is deleted and also when the partition doesn't have leadership information yet.